### PR TITLE
add summer editions 2023 feature flag to AppProvider FeaturesContext

### PR DESCRIPTION
**Pull request by gwyneplaine**
### WHY are these changes introduced?

Feature flag needed for rollout. 
Resolves https://github.com/Shopify/polaris-summer-editions/issues/33
Resolves https://github.com/Shopify/polaris-summer-editions/issues/32

### WHAT is this pull request doing?

* Add documented feature flag to AppProvider FeaturesContext. 
* Add explicit type (for better DX). 
* Add storybook story for QA
* Add addon-panel widget to storybook for toggling feature flag for stories. 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: polaris-react/src/utilities/features/types.ts [hunk 1]
```diff
@@ -1,5 +1,6 @@
 export interface FeaturesConfig {
-  [key: string]: boolean;
```
Comment (gwyneplaine) [inline polaris-react/src/utilities/features/types.ts:2]: explicit summer_editions boolean add to `FeaturesConfig` type for dx

File: polaris-react/tests/utilities/listbox.tsx [hunk 2]
```diff
@@ -26,10 +26,7 @@ export function mountWithListboxProvider(
     },
     render(element: React.ReactElement) {
       return (
-        <PolarisTestProvider
-          i18n={translations}
-          features={{newDesignLanguage: true}}
```
Comment (gwyneplaine) [inline polaris-react/tests/utilities/listbox.tsx:31]: not sure why this is still here since this was something we removed in v6 (afaik from the docs)

**Comment by gwyneplaine** [comment]
Notably @alex-page it appears the `useFeature` hook is not presently exposed as an export of `@shopify/polaris`. 
This is probably fine for 90% of our usecases for summer editions, but will mean any non polaris components that need access to the flag value won't be able to do so. If this is fine, we'll leave it. If its not fine, we can export it in this PR.

**Review by jesstelford — APPROVED** [review]

File: .changeset/khaki-eggs-shout.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'@shopify/polaris': patch
+---
+
+Add polaris_summer_editions_2023 feature flag to AppProvider context.
```
Comment (sophschneider) [inline .changeset/khaki-eggs-shout.md]: could we name this `polarisSummerEditions2023` instead? if not eslint in `Shopify/web` gets mad (`typescript-eslint/naming-convention`)

Screenshot 2023-05-17 at 4 35 50 PM

Comment (alex-page) [reply .changeset/khaki-eggs-shout.md]: Yes. The `underscore` approach is for the beta flag or feature flag.

File: polaris-react/src/utilities/features/types.ts [hunk 4]
```diff
@@ -1,5 +1,6 @@
...
+  polaris_summer_editions_2023?: boolean;
```
Comment (sophschneider) [inline polaris-react/src/utilities/features/types.ts]: see comment above in the changelog!
```suggestion
  polarisSummerEditions2023?: boolean;
```

File: polaris-react/src/components/AppProvider/AppProvider.tsx [hunk 5]
```diff
@@ -91,7 +91,12 @@ export class AppProvider extends Component<AppProviderProps, State> {
   };
 
   render() {
-    const {children, features = {}} = this.props;
+    const {
+      children,
+      features = {
+        polaris_summer_editions_2023: false,
```
Comment (sophschneider) [inline polaris-react/src/components/AppProvider/AppProvider.tsx]:
```suggestion
        polarisSummerEditions2023: false,
```

File: polaris-react/src/components/AppProvider/AppProvider.stories.tsx [hunk 6]
```diff
@@ -163,3 +165,44 @@ export function WithLinkComponent() {
     </AppProvider>
   );
 }
+
+export function WithSummerEditionsFeature() {
+  const [flagStatus, setFlagStatus] = useState(false);
+  const CheckFeature = ({children}: {children: React.ReactNode}) => {
+    const {polaris_summer_editions_2023} = useFeatures();
```
Comment (sophschneider) [inline polaris-react/src/components/AppProvider/AppProvider.stories.tsx]:
```suggestion
    const {polarisSummerEditions2023} = useFeatures();
```

File: polaris-react/.storybook/preview.js [hunk 7]
```diff
@@ -19,10 +19,15 @@ function StrictModeDecorator(Story, context) {
...
 
 function AppProviderDecorator(Story, context) {
+  const {polaris_summer_editions_2023} = context.globals;
   if (context.args.omitAppProvider) return <Story {...context} />;
-
   return (
-    <AppProvider i18n={enTranslations}>
+    <AppProvider
+      feature={{
+        polaris_summer_editions_2023,
```
Comment (sophschneider) [inline polaris-react/.storybook/preview.js]:
```suggestion
         polarisSummerEditions2023,
```

File: polaris-react/.storybook/preview.js [hunk 8]
```diff
@@ -19,10 +19,15 @@ function StrictModeDecorator(Story, context) {
 }
 
 function AppProviderDecorator(Story, context) {
+  const {polaris_summer_editions_2023} = context.globals;
```
Comment (sophschneider) [inline polaris-react/.storybook/preview.js]:
```suggestion
  const {polarisSummerEditions2023} = context.globals;
```

**Comment by sophschneider** [comment]
@gwyneplaine It should be fine for consumers to control the feature flag via the `AppProvider` features prop, I don't think they would need the `useFeature` hook since they're the one providing the boolean!

**Review by sophschneider — APPROVED** [review]
thanks for updating!

**Comment by sophschneider** [comment]
hmm I think the story is broken
Screenshot 2023-05-17 at 7 43 15 PM

**Review by alex-page — APPROVED** [review]

**Comment by gwyneplaine** [comment]
> hmm I think the story is broken Screenshot 2023-05-17 at 7 43 15 PM

AppProvider story isn't wrapped in an AppProvider, which is why that global toggle doesn't work. 
The point of that story is just a simple click test to make sure the feature flag is working 🎉

**Comment by sophschneider** [comment]
Thanks!!

**Comment by sophschneider** [comment]
Screenshot 2023-05-18 at 10 56 08 AM
Screenshot 2023-05-18 at 10 57 07 AM

hmm it doesnt work in any other story either. I think the `features` destructuring in the `AppProvider` is overriding everything

**Pull request merged**